circuit/types/integers/src/shl_checked.rs

```diff
@@ -68,27 +68,16 @@ impl<E: Environment, I: IntegerType, M: Magnitude> ShlChecked<Integer<E, M>> for
 
     #[inline]
     fn shl_checked(&self, rhs: &Integer<E, M>) -> Self::Output {
-        // Determine the variable mode.
-        if self.is_constant() && rhs.is_constant() {
-            // This cast is safe since `Magnitude`s can only be `u8`, `u16`, or `u32`.
-            match self.eject_value().checked_shl(rhs.eject_value().to_u32().unwrap()) {
-                Some(value) => Integer::new(Mode::Constant, console::Integer::new(value)),
-                None => E::halt("Constant shifted by constant exceeds the allowed bitwidth."),
-            }
-        } else {
-            // Determine the index where the first upper bit of the RHS must be zero.
-            // There is at least one trailing zero, as I::BITS = 8, 16, 32, 64, or 128.
-            let trailing_zeros_index = I::BITS.trailing_zeros() as usize;
-
-            // Determine if any of the upper bits of the RHS are nonzero.
-            let is_nonzero = rhs.bits_le[trailing_zeros_index..].iter().fold(Boolean::constant(false), |a, b| a | b);
-
-            // Ensure the upper bits of the RHS are zero.
-            E::assert(!is_nonzero);
+        let two = Self::one() + Self::one();
+        let base = match I::is_signed() {
+            // If `self` is a signed integer and is negative, then negate the base.
+            true => Self::ternary(&self.is_less_than(&Self::zero()), &(Self::zero() - &two), &two),
```

Comment:
If we introduce `is_negative` and `is_positive` trait methods, we can further reduce the cost of these `self.is_less_than(&Self::zero())` checks.

For signed cases, we can inspect that `self.msb().is_one()` when `I::is_signed()` for `is_negative`, and noop for `!I::is_signed()`. A similar case can be done for `is_positive`.

In Rust (and in math), 0 is neither positive nor negative. So we do need to perform a `.fold()` to ensure the case that all bits are 0 is `false` for `is_positive`.

Reply:
Agreed and thanks for the catch! Optimizing this implementation directly.